File: device-app/android/Java/helloWorld/app/src/main/java/com/rationalowl/hello/RoMessageListener.java
package com.rationalowl.hello;

import android.content.Context;
import android.content.Intent;
import android.util.Log;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.rationalowl.minerva.client.android.MessageListener;
import com.rationalowl.minerva.client.android.MinervaManager;
import com.rationalowl.minerva.client.android.util.Logger;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Map;

public class RoMessageListener implements MessageListener {
    private static final String TAG = "RoMsgListener";

    private static final ObjectMapper mapper = new ObjectMapper();


    @Override
    public void onMsgReceived(ArrayList<JSONObject> msgs) {
        Log.d(TAG, "onMsgReceived enter");

        int msgSize = msgs.size();

        Log.d(TAG, msgSize + " message received");

        try {
            JSONObject oneMsg = null;
            int msgType;
            String sender = null;
            long serverTime;
            String data = null;
            Map<String, String> customPush = null;

            // recent messages are ordered by message send time descending order [recentest, recent, old, older, ... oldest]
            for (JSONObject json : msgs) {
                // message type
                msgType = (int) json.get(MinervaManager.FIELD_MSG_TYPE);  // 1 (realtime: downstream), 2 (realtime: p2p), 3(custom push)
                // message sender (sender registration id)
                sender = (String) json.get(MinervaManager.FIELD_MSG_SENDER);
                // message send time
                serverTime = (Long) json.get(MinervaManager.FIELD_MSG_SERVER_TIME);
                // data
                data = (String) json.get(MinervaManager.FIELD_MSG_DATA);

                Log.d(TAG,  "message type:" + msgType);
                Log.d(TAG,  "message sender:" + sender);
                Log.d(TAG,  "message send time :" + serverTime);
                // custom data formatted json format
                Log.d(TAG,  "message :" + data);

                switch(msgType) {

                    case 1: // realtime downstream
                    case 2: {// realtime p2p
                        // this hello world don't handle realtime message
                        break;
                    }
                    // custom push received
                    case 3: {
                        // if multiple custom push received, we just notify recentest push only.
                        if(customPush == null) {
                            customPush = mapper.readValue(data, new TypeReference<Map<String, String>>() {});

                            String msgId = null, body = null, title = null, imageId = null;
                            // mandatory fields
                            msgId = customPush.get("mId");
                            body = customPush.get("body");

                            // optional fields
                            if(customPush.containsKey("title")) {
                                title = customPush.get("title");
                            }
                            if(customPush.containsKey("ii")) {
                                imageId = customPush.get("ii");
                            }

                            String[] extra = new String[4];
                            extra[0] = msgId;
                            extra[1] = body;
                            extra[2] = title;
                            extra[3] = imageId;
                            Context context = MinervaManager.getContext();
                            Intent intent = new Intent(MainActivity.MESSAGE_FROM_RATIONALOWL_MSG_LISTENER_ACTION);
                            intent.putExtra(MainActivity.MESSAGE_FROM_RATIONALOWL_MSG_LISTENER_KEY, extra);
                            context.sendBroadcast(intent);
                        }
                        break;
                    }
                    default:
                        break;
                }
            }

            if(customPush != null) {
                MyFirebaseMessagingService.showCustomNotification(customPush);
            }

        }
        catch (Exception e) {
            e.printStackTrace();
        }
    }


    @Override
    public void onSendUpstreamMsgResult(int resultCode, String resultMsg, String msgId) {
        Logger.debug(TAG, "onSendUpstreamMsgResult enter");
    }

    @Override
    public void onSendP2PMsgResult(int resultCode, String resultMsg, String msgId) {
        Logger.debug(TAG, "onSendP2PMsgResult enter");
    }
}

File: device-app/android/Kotlin/helloWorld/app/src/main/java/com/rationalowl/hello/RoMessageListener.kt
package com.rationalowl.hello

import android.content.Intent
import android.util.Log
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import com.rationalowl.minerva.client.android.MessageListener
import com.rationalowl.minerva.client.android.MinervaManager
import com.rationalowl.minerva.client.android.util.Logger
import org.json.JSONObject

class RoMessageListener : MessageListener {

    override fun onDownstreamMsgRecieved(msgList: java.util.ArrayList<JSONObject>) {
        // hello app don't treat realtime data
        Log.d(TAG, "onDownstreamMsgRecieved");
    }

    override fun onP2PMsgRecieved(msgList: java.util.ArrayList<JSONObject>) {
        // hello app don't treat realtime data
        Log.d(TAG, "onP2PMsgRecieved");
    }

    override fun onPushMsgRecieved(msgList: java.util.ArrayList<JSONObject>) {
        Log.d(TAG, "onPushMsgRecieved");
        val msgSize = msgList.size
        Log.d(TAG, "$msgSize message received")
        try {
            val oneMsg: JSONObject? = null

            var sender: String? = null
            var data: String? = null
            var customPush: Map<String, String>? = null

            // recent messages are ordered by message send time descending order [recentest, recent, old, older, ... oldest]
            for (json in msgList) {
                // message sender (sender registration id)
                sender = json[MinervaManager.FIELD_MSG_SENDER] as String
                // data
                data = json[MinervaManager.FIELD_MSG_DATA] as String
                Log.d(TAG, "message sender:$sender")
                // custom data formatted json format
                Log.d(TAG, "message :$data")

                // custom push format can be any fields you want.
                // RationalUms Demo format
                /*
                {
                    "mId": "message id here",
                    "title": "message title here",
                    "body": "message body here",
                    "ii": "image id here"
                    "st": "(message) send time"
                }
                */
                // if multiple custom push received, we just notify recentest push only.
                if (customPush == null) {

                    customPush = mapper.readValue(data)
                    var msgId: String? = null
                    var body: String? = null
                    var title: String? = null
                    var imageId: String? = null
                    // mandatory fields
                    msgId = customPush?.get("mId")
                    body = customPush?.get("body")

                    // optional fields
                    if (customPush!!.containsKey("title")) {
                        title = customPush?.get("title")
                    }
                    if (customPush.containsKey("ii")) {
                        imageId = customPush?.get("ii")
                    }
                    val extra = arrayOfNulls<String>(4)
                    extra[0] = msgId
                    extra[1] = body
                    extra[2] = title
                    extra[3] = imageId
                    val context = MinervaManager.getContext()
                    val intent =
                        Intent(MainActivity.Companion.MESSAGE_FROM_RATIONALOWL_MSG_LISTENER_ACTION)
                    intent.putExtra(
                        MainActivity.Companion.MESSAGE_FROM_RATIONALOWL_MSG_LISTENER_KEY,
                        extra
                    )
                    context.sendBroadcast(intent)
                }
            }
            if (customPush != null) {
                MyFirebaseMessagingService.Companion.showCustomNotification(customPush)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onSendUpstreamMsgResult(resultCode: Int, resultMsg: String, msgId: String) {
        Logger.debug(TAG, "onSendUpstreamMsgResult enter");
    }

    override fun onSendP2PMsgResult(resultCode: Int, resultMsg: String, msgId: String) {
        Logger.debug(TAG, "onSendP2PMsgResult enter");
    }

    companion object {
        private const val TAG = "RoMsgListener";
        private val mapper = jacksonObjectMapper();
    }
}